app/src/lib/stores/app-store.ts

```diff
@@ -1751,6 +1753,35 @@ export class AppStore extends TypedBaseStore<IAppState> {
     this._updateCurrentPullRequest(repository)
     this.updateMenuItemLabels(repository)
     this._initializeCompare(repository)
+    this.repositories = await this.getRefreshedRepositories(
+      this.repositories,
+      repository
+    )
+  }
+
+  public async refreshAllRepositories() {
+    this.repositories = await this.getRefreshedRepositories(this.repositories)
+    this.emitUpdate()
+  }
+
+  private async getRefreshedRepositories(
+    repositories: ReadonlyArray<Repository>,
+    targetRepo?: Repository
+  ): Promise<ReadonlyArray<Repository>> {
+    for (const repo of repositories) {
+      if (!targetRepo || targetRepo.id === repo.id) {
```

Comment:
🤔I can see two related use cases here - refreshing the local state of all repositories, or just refreshing one. Rather than doing a check like this, could we break this out into two functions?

```ts
  private async getRefreshedRepositories(
    repositories: ReadonlyArray<Repository>
  ): Promise<ReadonlyArray<Repository>> {
    for (const repo of repositories) {
      this.refreshRepositoryStatus(repo)
    }
    return repositories
  }

  private async refreshRepositoryStatus(
    repository: Repository
  ): Promise<Repository> {
    // do the actual refresh work here
  }
```

I think this also requires me to wrap my head around how we're currently managing `this.repositories` because we now have three places where we're updating it.

For existing changes to a `Repository` model we use the `RepositoriesStore` to update the database, like this:

https://github.com/desktop/desktop/blob/12af6e4bee8b398ea957ac7f51fdb07cc234946d/app/src/lib/stores/repositories-store.ts#L170-L173

This function then emits an update and the `AppStore` then re-reads what's in the database by invoking `getAll` this through the callback defined in `this.repositoriesStore.onDidUpdate`. I like this pattern because all changes flow the same way, but it doesn't feel like a natural fit for you because these changes don't need to be stored in the database.

Reply:
Hrmm perhaps I am not understanding Typescript's `ReadonlyArray` properly -- but I thought `this.repositories` being `readonly` means we cannot selectively update only one of the repositories, right? We will have to iterate through all of them to find that one and reassign the entire array anyway, hence my decision to add a `targetRepo` for when we only want to refresh one in hope for _some_ performance gain. 

I originally wanted to use `RepositoriesStore` to do all the updating as well - it just looked nicer to me, but yes, I ended up deciding against it for the reasons you cited. :)

Reply:
> Hrmm perhaps I am not understanding Typescript's `ReadonlyArray` properly -- but I thought `this.repositories` being `readonly` means we cannot selectively update only one of the repositories, right?

There's a couple of different uses for `readonly` in here, so I'll mention them here and then try and clarify my thoughts further:

 - `ReadonlyArray` is a way to pass around an array of values that won't be mutated by callers
 - the `readonly` keyword in TypeScript is a way to mark properties in a class which shouldn't be reassigned outside the constructor
 - `this.repositories` isn't defined as `readonly` inside `AppStore` so we can reassign it when the `RepositoriesStore` updates:

```ts
  private repositories: ReadonlyArray<Repository> = new Array<Repository>()
```

So while we're using a `ReadonlyArray<Repository>` in `getRefreshedRepositories` to indicate we're not modifying the array, we are mutating each repository in the array to update it's values and then returning the same array, which is then reassigned to `this.repositories`.